### ISSUE TITLE
Misinterpretation of highway=pedestrian oneway=yes

### ISSUE THREAD

**Issue by osmuser63783** [valhalla/valhalla#4014]
Valhalla won't route pedestrians against the flow of traffic in a pedestrian area (highway=pedestrian) that's been tagged oneway=yes. Or at least that's what I think is going on in this [example](https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=-3.205926418304444%2C55.950971946800266%2C-3.205668926239014%2C55.95125729879664).

Is this intentional? I think it's likely to be a misinterpretation of the oneway=yes tag which is [only intended to apply to vehicles](https://community.openstreetmap.org/t/effect-of-oneway-on-pedestrians/7631/19).

For comparison, OSRM and GraphHopper have no problems with sending pedestrians the "wrong way down a one way street."

**Comment by kevinkreiser** [comment]
@gknisely just mentioned this to me in passing the other day. i only looked into very very briefly but things like this:

https://github.com/valhalla/valhalla/blob/98fa75242d1853dcf2a58e9f29b775cc5d4dedc2/lua/graph.lua#L1182-L1192

are pretty unintuitive. i need time to review the access logic pattern we have going there, the code is so large its hard to look at (especially without comments) and quickly make sense of it

**Comment by nilsnolde** [comment]
Hm, not sure how to feel about this. It's clearly stated that `oneway` has to be ineffective for pedestrians, but clearly this way is only accessible for pedestrians! So what else would the oneway tag refer to if not to pedestrians? IMO this is a tagging error.

**Comment by osmuser63783** [comment]
[Streetview image](https://www.google.co.uk/maps/@55.9517223,-3.2032529,3a,34.3y,271.61h,89.77t/data=!3m6!1e1!3m4!1s9Kt9ivQwGkRPfQ0cfQfBAw!2e0!7i13312!8i6656) of how it's signed. It's a pedestrian zone, except for certain vehicles and times, and it's got a one way sign which is clearly only meant to apply to vehicles.

The tagging may be incomplete (it's missing the conditional exemption) but I don't think the combination of highway=pedestrian and oneway=yes is wrong. I've had a quick look and [found examples](https://overpass-turbo.eu/s/1sd0) in all the major European cities I checked.

**Comment by nilsnolde** [comment]
> It's a pedestrian zone, except for certain vehicles and times, 

Alright, I'm not exactly sure how best practices are in that case, but then it's missing some tagging IMO to indicate it's fine to use it conditionally or only as "destination". I can't confirm without a deeper look but I'd guess that Valhalla would respect that also for pedestrian zones and give vehicles access conditionally. It doesn't solve the `oneway=yes` tag issue, but now I see your point, if it's accessible to vehicles a `oneway=yes` should also only apply to vehicles, and pedestrian should only be one-way if it's `oneway:foot=yes`. And generally we could just avoid using `oneway` for pedestrian. So, the missing tagging really has nothing to do with it, but for my own understanding it'd have been more obvious:)

BTW, I don't see an indication on the Google Streetview image that the road is a one-way, right? It might be used like that, but doesn't seem to be signed like that, unless Scotland works a bit differently than Germany.

**Comment by osmuser63783** [comment]
The one way sign is the blue arrow pointing up, it's on the wall to the right.

On the other end there will be a no entry sign. You can see one if you turn the camera by 180 degrees.

**Comment by gknisely** [comment]
Seem like [this](https://www.openstreetmap.org/way/4237322#map=18/55.95114/-3.20394) should be tagged with the following:
`highway=pedestrian`
`motor_vehicle:conditional=delivery @ (23:00-10:30)`
`oneway=yes`

However, I do think there is a bug here as we shut off all access for vehicles if we encounter `highway=pedestrian`  Also, we need to be able to handle cases where it is truly one-way exits as shown [here](https://www.openstreetmap.org/way/806874233)  Regardless, this is a bug for sure.

**Comment by osmuser63783** [comment]
Yes, it seems like the true one way for pedestrians example would more accurately be tagged oneway:foot=yes, or foot:backward=yes which was the favoured option when this was [discussed](https://community.openstreetmap.org/t/effect-of-oneway-on-pedestrians/7631/20?u=osmuser63783) in the community forum.

There are 26k cases of highway=pedestrian and 44k cases of highway=footway with oneway=* tags. The question is, how often are these meant to indicate actual one-way pedestrian access, so that it would be a problem to route pedestrians both ways regardless of access restrictions? I would think very rarely. Whereas if you assume that they are oneway unless specific other tags are present, then it's easy to make mistakes.

The access restriction could be unmapped, like in my example. But checking for access restrictions isn't enough. You would also need to check for bicycle=yes. highway=footway bicycle=yes oneway=yes is legitimate tagging for a footway=sidewalk where it's common that bikes are only allowed to go one way but pedestrians can of course walk both ways.

**Comment by nilsnolde** [comment]
Uff haha yeah that all exists I guess. Again, I’m also in favor of never allowing oneway for pedestrians other than it being explicitly tagged lilt that. I think the wiki is also quite clear about it. 

So, in case of a oneway on pedestrian-only ways (like your example), I guess we’d just assume that any other mode is oneway, whichever has access to that way, e.g. bike, but pedestrian is both ways. And then we’d consider @gknisely ‘s example to be wrongly tagged (ie it should have explicit „oneway:foot=true“)? That’d be preferred from my side and sounds a bit easier on the lua side (without looking in depth).

Opinions?

**Comment by osmuser63783** [comment]
This seems sensible to me.. though I'm just a user who stumbled upon a bug. The oneway tag would just be ignored altogether in pedestrian routing. Only oneway:foot, foot:backward and foot:forward should affect it.

**Comment by kadiwa4** [comment]
Also applies to `highway=path`'s for both bike and pedestrian use: [example](https://www.openstreetmap.org/directions?engine=fossgis_valhalla_foot&route=53.59271%2C9.98827%3B53.59434%2C9.98806#map=18/53.59358/9.98778).

**Comment by osmuser63783** [comment]
The above example has since been edited to `footway`. Here are some examples for `path`: [1](https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=9.735099077224733%2C52.39456322844031%2C9.735753536224367%2C52.39913276574284), [2](https://www.openstreetmap.org/directions?engine=fossgis_valhalla_foot&route=49.49090%2C11.05423%3B49.48924%2C11.05552#map=18/49.48993/11.05563).

Community consensus seems to be, also for `path`, that `oneway` is only supposed to apply to bicycles: [forum thread today](https://community.openstreetmap.org/t/how-to-map-one-way-cycle-foot-path-where-only-the-cycleway-is-one-way/109695).

**Comment by nilsnolde** [comment]
closing here in favor of https://github.com/valhalla/valhalla/issues/4978
